lib/Target/CBackend/CBackend.cpp

```diff
@@ -4600,49 +4606,68 @@ void CWriter::printIntrinsicDefinition(FunctionType *funT, unsigned Opcode,
 
     case Intrinsic::bswap:
       cwriter_assert(retT == elemT);
-      Out << "  LLVMFlipAllBits(8 * sizeof(a), &a, &r);\n";
+      cwriter_assert(!isa<VectorType>(retT));
+      cwriter_assert(elemT->getIntegerBitWidth() <= 64);
+      Out << "  int i;\n";
+      Out << "  r = 0;\n";
+      Out << "  int bitwidth = " << elemT->getIntegerBitWidth() << ";\n";
+      Out << "  for (i = 0; i < bitwidth/8; i++)\n";
+      Out << "    r |= ((a >> (i*8)) & 0xff) << (bitwidth - (i+1)*8);\n";
       break;
 
     case Intrinsic::ctpop:
       cwriter_assert(retT == elemT);
-      Out << "  r = ";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << "llvm_ctor_u128(0, ";
-      Out << "LLVMCountPopulation(8 * sizeof(a), &a)";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << ")";
-      Out << ";\n";
+      cwriter_assert(!isa<VectorType>(retT));
+      cwriter_assert(elemT->getIntegerBitWidth() <= 64);
+      Out << "  int i;\n";
+      Out << "  r = 0;\n";
+      Out << "  int bitwidth = " << elemT->getIntegerBitWidth() << ";\n";
+      Out << "  for (i = 0; i < bitwidth; i++)\n";
+      Out << "    if ( (a >> i ) & 1 )\n";
+      Out << "      r++;\n";
+
       break;
 
     case Intrinsic::ctlz:
       cwriter_assert(retT == elemT);
-      Out << "  (void)b;\n  r = ";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << "llvm_ctor_u128(0, ";
-      Out << "LLVMCountLeadingZeros(8 * sizeof(a), &a)";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << ")";
-      Out << ";\n";
+      cwriter_assert(!isa<VectorType>(retT));
+      cwriter_assert(elemT->getIntegerBitWidth() <= 64);
+      Out << "  int i;\n";
+      Out << "  r = 0;\n";
+      Out << "  int bitwidth = " << elemT->getIntegerBitWidth() << ";\n";
+      Out << "  for (i = bitwidth - 1; i >= 0; i--)\n";
+      Out << "    if ( ((a >> i ) & 1) == 0 )\n";
+      Out << "      r++;\n";
+      Out << "    else\n";
+      Out << "      break;\n";
       break;
 
     case Intrinsic::cttz:
       cwriter_assert(retT == elemT);
-      Out << "  (void)b;\n  r = ";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << "llvm_ctor_u128(0, ";
-      Out << "LLVMCountTrailingZeros(8 * sizeof(a), &a)";
-      if (retT->getPrimitiveSizeInBits() > 64)
-        Out << ")";
-      Out << ";\n";
+      cwriter_assert(!isa<VectorType>(retT));
+      cwriter_assert(elemT->getIntegerBitWidth() <= 64);
+      Out << "  int i;\n";
+      Out << "  r = 0;\n";
+      Out << "  int bitwidth = " << elemT->getIntegerBitWidth() << ";\n";
+      Out << "  for (i = 0; i < bitwidth; i++)\n";
+      Out << "    if ( ((a >> i) & 1) == 0 )\n";
+      Out << "      r++;\n";
+      Out << "    else\n";
+      Out << "      break;\n";
       break;
 
     case Intrinsic::umax:
+    case Intrinsic::smax:
       Out << "  r = a > b ? a : b;\n";
```

Comment:
This might need a type cast (to make it a signed comparison)

Reply:
I don't think this is necessary in this case: the types of `a` and `b` should already be signed in this case (through https://github.com/JuliaHubOSS/llvm-cbe/pull/188/files/53cacb77ab80bb98ef3bc620f603bcfea136babc#diff-18ffe28a234c38c6d09cd7a1e882d3bfa7a18faf6d119346b5417d476619461dR4491-R4493) which should result in the comparison being signed?